src/main/java/wooteco/subway/domain/Sections.java

```diff
@@ -0,0 +1,215 @@
+        return collection.stream()
+            .map(mapper)
+            .collect(Collectors.toList());
+    }
+
+    private <T> List<T> getSortedValuesBy(Long firstKey, Function<Long, Long> newKeyMapper, Map<Long, T> values) {
+        List<T> result = new ArrayList<>();
+        Long key = firstKey;
+
+        while (values.containsKey(key)) {
+            result.add(values.get(key));
+            key = newKeyMapper.apply(key);
+        }
+
+        return result;
+    }
+
+    public Section add(Section newSection) {
+        validateAddable(newSection);
+
+        if (getUpDestination().equals(newSection.getDownStation())) {
+            values.add(0, newSection);
+            return newSection;
+        }
+
+        if (getDownDestination().equals(newSection.getUpStation())) {
+            values.add(newSection);
+            return newSection;
+        }
+
+        for (Section section : values) {
+            if (section.hasSameUpStation(newSection)) {
+                return insertAfterUpStation(newSection, section);
+            }
+
+            if (section.hasSameDownStation(newSection)) {
+                return insertBeforeDownStation(newSection, section);
+            }
+        }
+        throw new IllegalStateException("구간을 추가할 수 있는 상태가 아닙니다.");
+    }
+
+    private Section insertAfterUpStation(Section newSection, Section section) {
+        int index = values.indexOf(section);
+        Section sectionToModify = new Section(section.getId(), newSection.getDownStation(), section.getDownStation(),
+            getNewDistance(section, newSection));
+        values.set(index, newSection);
+        values.add(index + 1, sectionToModify);
+        return sectionToModify;
+    }
+
+    private int getNewDistance(Section section, Section newSection) {
+        int newDistance = section.getDistance() - newSection.getDistance();
+        if (newDistance <= 0) {
+            throw new IllegalArgumentException("추가하려는 구간의 거리가 기존 구간의 거리보다 작아야합니다.");
+        }
+        return newDistance;
+    }
+
+    private Section insertBeforeDownStation(Section newSection, Section section) {
+        int index = values.indexOf(section);
+        Section sectionToModify = new Section(section.getId(), section.getUpStation(), newSection.getUpStation(),
+            getNewDistance(section, newSection));
+        values.set(index, sectionToModify);
+        values.add(index + 1, newSection);
+        return sectionToModify;
+    }
+
+    private void validateAddable(Section section) {
+        List<Section> foundSections = findSectionsOverlapped(section);
+
+        if (isInvalid(section, foundSections)) {
+            throw new IllegalArgumentException("이미 포함된 두 역을 가진 Section 을 추가할 수 없습니다.");
+        }
+    }
+
+    private List<Section> findSectionsOverlapped(Section newSection) {
+        return values.stream()
+            .filter(section ->
+                section.contains(newSection.getDownStation()) || section.contains(newSection.getUpStation()))
```

Comment:
이것도 `Section`이 메서드로 제공해 줄 수 있을 것 같아 보여요.

Reply:
section 내에서 비교 로직을 수행하도록 다음과 같이 변경해보았습니다. 감사합니다👍👍

```suggestion
                section -> section.isOverlap(newSection)
```